PyPortal_Wakeup_Light/wake_up_light.py

```diff
@@ -0,0 +1,198 @@
+        if current_hour < wake_up_hour and current_minutes < wake_up_minutes:
+            if current_day == day:
+                input_wake_up_time = wake_up_times[day]
+    # set wake up time to the next day's wake up time the night before
+        else:
+            if current_day == 6:
+                input_wake_up_time = wake_up_times[0]
+            else:
+                if current_day == day:
+                    input_wake_up_time = wake_up_times[day+1]
+    input_wake_up_time_text = "Wake up at " + input_wake_up_time
+    wakeup_time_textarea.text = input_wake_up_time_text
+    return input_wake_up_time
+
+def subtract30min(time_before):
+    hours_before, minutes_before = time_before.split(":")
+    AM_PM_str = minutes_before[-1:]
+    minutes_before = int(minutes_before[:-1])
+    if AM_PM_str == 'P':
+        hours_before = int(hours_before) + 12
+    elif ((hours_before == '12') and (AM_PM_str == 'A')):
+        hours_before = 0
+    else:
+        hours_before = int(hours_before)
+    if  minutes_before >= 30:
+        minutes_after = minutes_before - 30
+        format_str = "%d:%02d"
+        if hours_before >= 12:
+            hours_after = hours_before - 12
+            format_str = format_str+"P"
+        else:
+            hours_after = hours_before
+            format_str = format_str+"A"
+        if hours_before == 0:
+            hours_after = 12
+        sub30_str = format_str % (hours_after, minutes_after)
+    elif minutes_before < 30:
+        minutes_after = minutes_before + 30
+        hours_after = hours_before - 1
+        format_str = "%d:%02d"
+        if hours_before >= 12:
+            hours_after = hours_after - 12
+            format_str = format_str+"P"
+        else:
+            format_str = format_str+"A"
+        if hours_before == 0:
+            hours_after = 11
+            format_str = format_str[:-1]+"P"
+        sub30_str = format_str % (hours_after, minutes_after)
+    light_on_time_textarea.text = "Light starting at: " + sub30_str
+    return sub30_str
+
+def displayTime():
+    now = time.localtime()
+    hour, minute = now[3:5]
+    print(now)
+    print("Current time: %02d:%02d" % (hour, minute))
+    # display the time in a nice big font
+    format_str = "%d:%02d"
+    if hour >= 12:
+        hour -= 12
+        format_str = format_str+"P"
+    else:
+        format_str = format_str+"A"
+    if hour == 0:
+        hour = 12
+    time_str = format_str % (hour, minute)
+    time_textarea.text = time_str
+    return time_str
+
+refresh_time = None
+
+while True:
+    # only query the online time once per hour (and on first run)
+    if (not refresh_time) or (time.monotonic() - refresh_time) > 3600:
+        try:
+            print("Getting time from internet!")
+            pyportal.get_local_time()
+            refresh_time = time.monotonic()
+        except RuntimeError as e:
+            print("Some error occured, retrying! -", e)
+            continue
+    time_str_text = displayTime()
```

Comment:
use time objects and math, dont store/display the time with PM or whatever, unless its to the screen

Reply:
For this do you want the user to input their wake up times as 24 hour time values? Reason I did it this way is so user can enter in their wake up times in 12 hour system (AM and PM).

Reply:
if you want them to enter wakeup time with A/P then parse that time *once* at the beginning of the program. and immediately convert it to a time object. you definitely should not be parsing/unparsing strings of time. makes the code clunkier than it needs to be

Reply:
Ah that makes sense. Ok will add those changes!